### ISSUE TITLE
Using HistogramIterationListener blocks usage of the GPU.

### ISSUE THREAD

**Issue by wmeddie** [deeplearning4j/deeplearning4j#2039]
Same setup as in #2037

![nohistogram](https://cloud.githubusercontent.com/assets/122480/18027129/2a4fc53a-6c96-11e6-924b-605b3daeb5e8.png)
![histogram](https://cloud.githubusercontent.com/assets/122480/18027140/65faa582-6c96-11e6-818d-83d28d8b1109.png)

Without a HistogramIterationListener:

```
o.d.o.l.PerformanceListener - iteration 1; iteration time: 458 ms; samples/sec: 139.738; batches/sec: 2.183; 
o.d.o.l.PerformanceListener - iteration 2; iteration time: 56 ms; samples/sec: 1142.857; batches/sec: 17.857; 
o.d.o.l.PerformanceListener - iteration 3; iteration time: 70 ms; samples/sec: 914.286; batches/sec: 14.286; 
o.d.o.l.PerformanceListener - iteration 4; iteration time: 43 ms; samples/sec: 1488.372; batches/sec: 23.256; 
o.d.o.l.PerformanceListener - iteration 5; iteration time: 76 ms; samples/sec: 842.105; batches/sec: 13.158; 
o.d.o.l.PerformanceListener - iteration 6; iteration time: 31 ms; samples/sec: 2064.516; batches/sec: 32.258; 
o.d.o.l.PerformanceListener - iteration 7; iteration time: 81 ms; samples/sec: 790.123; batches/sec: 12.346; 
o.d.o.l.PerformanceListener - iteration 8; iteration time: 42 ms; samples/sec: 1523.810; batches/sec: 23.810; 
o.d.o.l.PerformanceListener - iteration 9; iteration time: 33 ms; samples/sec: 1939.394; batches/sec: 30.303; 
o.d.o.l.PerformanceListener - iteration 10; iteration time: 47 ms; samples/sec: 1361.702; batches/sec: 21.277; 
```

With a HistogramIterationLister:

```
INFO  [2016-08-27 11:27:35,511] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 1; iteration time: 1843 ms; samples/sec: 34.726; batches/sec: 0.543; 
INFO  [2016-08-27 11:27:36,818] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 2; iteration time: 1307 ms; samples/sec: 48.967; batches/sec: 0.765; 
INFO  [2016-08-27 11:27:37,873] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 3; iteration time: 1055 ms; samples/sec: 60.664; batches/sec: 0.948; 
INFO  [2016-08-27 11:27:38,877] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 4; iteration time: 1004 ms; samples/sec: 63.745; batches/sec: 0.996; 
INFO  [2016-08-27 11:27:39,875] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 5; iteration time: 998 ms; samples/sec: 64.128; batches/sec: 1.002; 
INFO  [2016-08-27 11:27:40,894] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 6; iteration time: 1019 ms; samples/sec: 62.807; batches/sec: 0.981; 
INFO  [2016-08-27 11:27:41,877] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 7; iteration time: 983 ms; samples/sec: 65.107; batches/sec: 1.017; 
INFO  [2016-08-27 11:27:42,945] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 8; iteration time: 1068 ms; samples/sec: 59.925; batches/sec: 0.936; 
INFO  [2016-08-27 11:27:44,015] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 9; iteration time: 1070 ms; samples/sec: 59.813; batches/sec: 0.935; 
INFO  [2016-08-27 11:27:45,046] org.deeplearning4j.optimize.listeners.PerformanceListener: iteration 10; iteration time: 1031 ms; samples/sec: 62.076; batches/sec: 0.970; 
```

**Comment by raver119** [comment]
Well, it's not really a news tbh. HistogramListener actually reprocesses your model params, submits them to the webserver, waits for response. And that's happening on each iteration, according to your frequency configuration. 

So i think it's more like "enhancement request", then bug.

**Comment by wmeddie** [comment]
Understandable.  I wasn't expecting such a drastic performance drop though.  Seems that upping the iterations to HistogramIterationListener has no effect in 0.5.0.  The decompiler shows that it just sets it to 1.  Looks like it's fixed in master.

**Comment by anovstrup** [comment]
Interesting. I experienced poor performance (i.e., no noticeable difference) when switching from CPU to GPU recently. It may just be that the GPU on my laptop isn't great, but I was also using `HistogramIterationListener`.

**Comment by raver119** [comment]
Well, as i've said before - it's known issue. And even more - it was solved once. But unfortunately commit was partially lost :(

So in few nearest days i'm going to solve this issue once again, by moving HTTP communication into background thread. That's why this issue isn't closed yet :)

**Comment by ramgo2** [comment]
I see two possible scenarios where this is useful..
1. Continuously watching the UI to get early feedback on how the network is training.  Abort runs if not in line with expectation.
2. Let the training complete.  Visit the graph to see how the parameters affected the training.
While in both cases performance is important, the first scenario may be less affected by slowness as probably one may not sit and watch for a long time or the volume of training data may be small.  However, for the second scenario it really matters whether training completes in 8 hours or 48 hours (I experience 6 times increase in running time with HistorgramIterationListener).  Can an option be provided where only the data required for the graphs is logged without a live UI (if that can improve performance)?

**Comment by raver119** [comment]
we're in the process of solving this issue.

**Comment by wmeddie** [comment]
This was fixed with 0.6.0.
